### DISCUSSION TITLE
GCN_Norm for directed graph

### DISCUSSION THREAD

**Discussion by qubit0** [pyg-team/pytorch_geometric#5445] [Q&A] [answered]
`WebKB` dataset has a directed graph and also has a few self-loop. How do we use `gcn_norm` here? What is the difference between `source_to_target` and `target_to_source` ? While adding a self-loop, does `gcn_norm` check if there is already a self-loop or not (It seems like `add_remaining_self_loop` is already doing that)?

 If I use a naive implementation of GCN_norm (for example this [one ](https://github.com/danielegrattarola/spektral/blob/master/spektral/utils/convolution.py)) then it will just add an extra self-loop (even when there is already a self-loop) and also uses `target_to_source`like calculation.

**Comment by rusty1s** [comment, marked as answer]
In theory, `gcn_norm` is only well-defined on undirected graphs since it takes the degrees for both source and destination node into account. Ofc, it will still give you a reasonable output for directed graphs, but you may need to choose whether to do the computation based on the in-degree (`target_to_source`) or out-degree (`source_to_target`). `gcn_norm` will check if there is already a self-loop present and will not another one for these nodes.
